### PR TITLE
[lldb][DWARFIndex] Use IDX_parent to implement GetFullyQualifiedType query

### PR DESCRIPTION
**Pull request by felipepiovezan**
This commit changes DebugNamesDWARFIndex so that it now overrides `GetFullyQualifiedType` and attempts to use DW_IDX_parent, when available, to speed up such queries. When this type of information is not available, the base-class implementation is used.

With this commit, we now achieve the 4x speedups reported in [1].

[1]: https://discourse.llvm.org/t/rfc-improve-dwarf-5-debug-names-type-lookup-parsing-speed/74151/38

### CODE REVIEW

**Comment by felipepiovezan** [comment]
The unittests here depend on https://github.com/llvm/llvm-project/pull/79666

File: lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp
Comment (clayborg) [inline lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:291]: Maybe use less continues here?
```
if (parent_chain) {
  if (SameParentChain(parent_names, *parent_chain) &&
      !ProcessEntry(entry, callback))
    return;
} else {
  if (!fallback_impl(entry))
    return;
}
```

File: lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp
Comment (adrian-prantl) [inline lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:230]: isn't `namespace` and `static` redundant?

File: lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp
Comment (adrian-prantl) [inline lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:239]: comment should be on its own line.

File: lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp
Comment (adrian-prantl) [inline lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:289]: This might benefit fro a few high-level comments about what's happening in this loop?

File: lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp
Comment (adrian-prantl) [inline lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:295]: Any comments that would be useful in this function?

File: lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp
Comment (felipepiovezan) [reply lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:230]: Ooops.
Errr let me ask the specialist:
image

File: lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp
Comment (dwblaikie) [inline lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:267]: Not sure this is worth putting in a lambda? Might be simpler to write it inline in the one place it's used?

File: lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp
Comment (dwblaikie) [reply lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:291]: FWIW as-written does lean a bit towards the LLVM documented style ( https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code ) - but I wouldn't be too fussed either way.

File: lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp
Comment (dwblaikie) [inline lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:312]: Similarly, a lambda that's only called once doesn't seem to carry its weight - might be simpler to put it in the loop.

File: lldb/unittests/SymbolFile/DWARF/DWARFDebugNamesIndexTest.cpp
Comment (dwblaikie) [inline lldb/unittests/SymbolFile/DWARF/DWARFDebugNamesIndexTest.cpp:32]: Probably inline the lambda at the call site?
```
index.GetFullyQualifiedType(ctx, [*](DWARFDIE die) {
  num_matches++;
  return true;
});
```

File: lldb/unittests/SymbolFile/DWARF/DWARFDebugNamesIndexTest.cpp
Comment (dwblaikie) [inline lldb/unittests/SymbolFile/DWARF/DWARFDebugNamesIndexTest.cpp:206]: Since it's written twice, might be OK to have this as a file-static helper function rather than two separate copies as lambdas?

File: lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp
Comment (felipepiovezan) [reply lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:312]: I'm not attached to the lambda, but I would like to clarify my original intent with the lambda and then double check if you still prefer the inlined version with that intent in mind.

My intent was not to re-use code, but rather to give name to things and reduce cognitive burden. If we go with the suggestion, this is what we end up with:

```
  for (auto [parent_name, parent_entry] :
        llvm::zip_equal(parent_names, parent_entries))
    auto maybe_dieoffset = entry.getDIEUnitOffset();
     if (!maybe_dieoffset)
       return false;
     auto die_ref = ToDIERef(entry);
     if (!die_ref)
       return false;
     return name == m_debug_info.PeekDIEName(*die_ref);
```

A reader needs to go through that entire block and conclude: "Oh, this is just comparing the AT_name with `name`.

With the current version, the code _tells_ you what it is doing 
`if (!SameAsEntryATName(parent_name, parent_entry))`

With that in mind, if you still prefer the inlined version, I'll be happy to update it!

(the same rationale applies to the other comment about lambdas)

File: lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp
Comment (felipepiovezan) [reply lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:291]: If that's ok, I propose we leave it as is because of the guidelines and because I can imagine this method growing in the future to include more code after the referenced line.

File: lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp
Comment (felipepiovezan) [reply lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:289]: I'll add something at the top of loop

**Comment by felipepiovezan** [comment]
Addressed all review comments (waiting for a reply on the one comment about inlining a lambda)

File: lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp
Comment (felipepiovezan) [reply lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp]: I've inlined this one, please have a look

File: lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp
Comment (adrian-prantl) [inline lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:241]: Should/could we LLDB_LOG this to the dwarf channel here?

**Review by adrian-prantl — APPROVED** [review]

File: lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp
Comment (clayborg) [inline lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:277]: Since we are only fetching the same number of items, does this mean that a search for "a::b::c" will match "z::a::b::c" if we only fetch 2 parent items? It seems if the parent chain size differs then we want to not match anything?

Comment (felipepiovezan) [reply lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:277]: Not sure I follow, the intent is to only match things that are 100% identical.
The parent chain size is checked first thing inside `SameParentChain`:

```
bool DebugNamesDWARFIndex::SameParentChain(...) {
  if (parent_entries.size() != parent_names.size())
     return false;
```

Comment (felipepiovezan) [reply lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:277]: The `parent_names.size() + 1` is only an optimization (we only need one extra parent to determine that the parent scopes have a different number of entries)

Comment (felipepiovezan) [reply lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:277]: > Since we are only fetching the same number of items

Actually, rereading this portion of your comment, this is the faulty assumption. We are not fetching the same number of items, we are fetching +1 items.

File: lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp
Comment (felipepiovezan) [reply lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:241]: Good point, let me look into it

**Comment by felipepiovezan** [comment]
@adrian-prantl I've added a log message with the `Error` itself being logged as well. That said, printing any information specific to the entry itself would require parsing other Form and, given that we already failed to parse one, I figured it would be better not to attempt parsing more while logging.

**Comment by adrian-prantl** [comment]
> I've added a log message with the `Error` itself being logged as well

Thanks! I wasted half a day stepping through the DWARF parser recently, thinking I'm debugging a parser bug, when it was really invalid DWARF produced  by an older version of dsymutil, so even just the information that there _was_ a parse error is very useful.

**Comment by clayborg** [comment]
> > I've added a log message with the `Error` itself being logged as well
> 
> Thanks! I wasted half a day stepping through the DWARF parser recently, thinking I'm debugging a parser bug, when it was really invalid DWARF produced by an older version of dsymutil, so even just the information that there _was_ a parse error is very useful.

I always run `llvm-dwawfdump --verify` as a first step, though it is nice to get more errors emitted in the DWARF log channels

**Comment by felipepiovezan** [comment]
Any other comments here? I would like to land this once the YAML review is unblocked

**Comment by felipepiovezan** [comment]
I will aim to land this sometime tomorrow if there are no further objections

**Comment by dwblaikie** [comment]
> I will aim to land this sometime tomorrow if there are no further objections

Per the documentation ( https://llvm.org/docs/CodeReview.html#code-review-workflow ), please don't do this. Once it's sent for review, please wait for approval (ping as needed, etc) before landing.

**Comment by jimingham** [comment]
> > I will aim to land this sometime tomorrow if there are no further objections
> 
> Per the documentation ( https://llvm.org/docs/CodeReview.html#code-review-workflow ), please don't do this. Once it's sent for review, please wait for approval (ping as needed, etc) before landing.

Felipe already had Adrian's approval, and so far as I can see there were no outstanding comments he hadn't addressed.  Surely in that case, you don't have to wait for all the other people selected as reviewers to weigh in do you?  That would be pretty unworkable, lots of people assigned as reviewers don't...

Seems to me since this PR already had an approval and no No's, Felipe could have committed it right away.  Holding on for late reviewers seems like a courtesy, not a violation of the "you must have an approval" policy.

**Comment by dwblaikie** [comment]
> > > I will aim to land this sometime tomorrow if there are no further objections
> > 
> > 
> > Per the documentation ( https://llvm.org/docs/CodeReview.html#code-review-workflow ), please don't do this. Once it's sent for review, please wait for approval (ping as needed, etc) before landing.
> 
> Felipe already had Adrian's approval, and so far as I can see there were no outstanding comments he hadn't addressed. 

Ah, sorry - I scanned up and down but missed it (the UI's not quite as explicit as Phab is on whether a patch has been reviewed - but I could've looked in the top right and seen the check next to Adrian's name).

Yeah, I don't expect multiple reviewer approval unless folks (either the submitter, or other reviewers) ask for it. If the it's approved with some optional comments and the discussion around those aspects has been resolved, fine by me to commit.

**Pull request merged**
